distributed/threadpoolexecutor.py

```diff
@@ -20,44 +20,75 @@
 
    Copyright 2001-2016 Python Software Foundation; All Rights Reserved
 """
-import itertools
 import logging
-import os
 import queue
 import threading
+import weakref
 
 from . import _concurrent_futures_thread as thread
-from .metrics import time
 
 logger = logging.getLogger(__name__)
 
 thread_state = threading.local()
 
 
-def _worker(executor, work_queue):
+def _worker_secedable(executor_reference, work_queue, initializer, initargs):
+    if initializer is not None:
+        try:
+            initializer(*initargs)
+        except BaseException:
+            logger.critical("Exception in initializer:", exc_info=True)
+            executor = executor_reference()
+            if executor is not None:
+                executor._initializer_failed()
+            return
     thread_state.proceed = True
-    thread_state.executor = executor
-
+    thread_state.executor = executor_reference()
     try:
         while thread_state.proceed:
+            executor = executor_reference()
+            if not executor:
+                return
             with executor._rejoin_lock:
                 if executor._rejoin_list:
                     rejoin_thread, rejoin_event = executor._rejoin_list.pop()
                     executor._threads.add(rejoin_thread)
-                    executor._threads.remove(threading.current_thread())
+                    executor._threads.discard(threading.current_thread())
                     rejoin_event.set()
+                    executor._idle_semaphore.acquire(timeout=0)
```

Comment:
This is where I'm not entirely sure if we're doing the right thing. If I do not acquire, we're deadlocking right away. Acquiring too often would cause the counter to underestimate idle threads and spawn more than necessary, therefore it should not be a huge deal

Reply:
deadlocks happen in `test_rejoin_idempotent`. I didn't have the patience to debug this any futher since I think it should not be a big deal.